src/lib.rs

```diff
@@ -3789,6 +3805,514 @@ pub struct SemanticHighlightingParams {
+    pub const DECLARATION: SemanticTokenModifier = SemanticTokenModifier::new("declaration");
+    pub const DEFINITION: SemanticTokenModifier = SemanticTokenModifier::new("definition");
+    pub const REFERENCE: SemanticTokenModifier = SemanticTokenModifier::new("reference");
+    pub const STATIC: SemanticTokenModifier = SemanticTokenModifier::new("static");
+    pub const ABSTRACT: SemanticTokenModifier = SemanticTokenModifier::new("abstract");
+    pub const DEPRECATED: SemanticTokenModifier = SemanticTokenModifier::new("deprecated");
+    pub const ASYNC: SemanticTokenModifier = SemanticTokenModifier::new("async");
+    pub const VOLATILE: SemanticTokenModifier = SemanticTokenModifier::new("volatile");
+    pub const READONLY: SemanticTokenModifier = SemanticTokenModifier::new("readonly");
+
+    pub const fn new(tag: &'static str) -> Self {
+        SemanticTokenModifier(Cow::Borrowed(tag))
+    }
+
+    pub fn as_str(&self) -> &str {
+        &self.0
+    }
+}
+
+#[cfg(feature = "proposed")]
+impl From<String> for SemanticTokenModifier {
+    fn from(from: String) -> Self {
+        SemanticTokenModifier(Cow::from(from))
+    }
+}
+#[cfg(feature = "proposed")]
+impl From<&'static str> for SemanticTokenModifier {
+    fn from(from: &'static str) -> Self {
+        SemanticTokenModifier::new(from)
+    }
+}
+
+/**
+ * @since 3.16.0 - Proposed state
+ */
+#[derive(Debug, Eq, PartialEq, Clone, Default, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+#[cfg(feature = "proposed")]
+pub struct SemanticTokensLegend {
+    /**
+     * The token types a server uses.
+     */
+    pub token_types: Vec<SemanticTokenType>,
+
+    /**
+     * The token modifiers a server uses.
+     */
+    pub token_modifiers: Vec<SemanticTokenModifier>,
+}
+
+/**
+ * The actual tokens. For a detailed description about how the data is
+ * structured please see
+ * https://github.com/microsoft/vscode-extension-samples/blob/5ae1f7787122812dcc84e37427ca90af5ee09f14/semantic-tokens-sample/vscode.proposed.d.ts#L71
+ */
+#[derive(Debug, Eq, PartialEq, Copy, Clone, Default)]
+#[cfg(feature = "proposed")]
+pub struct SemanticToken {
+    pub delta_line: u32,
+    pub delta_start: u32,
+    pub length: u32,
+    pub token_type: u32,
+    pub token_modifiers_bitset: u32,
+}
+
+#[cfg(feature = "proposed")]
+impl SemanticToken {
+    fn deserialize_tokens<'de, D>(deserializer: D) -> Result<Vec<SemanticToken>, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let data = Vec::<u32>::deserialize(deserializer)?;
+
+        let mut res = Vec::new();
+
+        for chunk in data.chunks_exact(5) {
+            res.push(SemanticToken {
+                delta_line: chunk[0],
+                delta_start: chunk[1],
+                length: chunk[2],
+                token_type: chunk[3],
+                token_modifiers_bitset: chunk[4],
+            })
+        }
+
```

Comment:
Let's check that the https://doc.rust-lang.org/std/slice/struct.ChunksExact.html#method.remainder is empty and emit errors otherwise.

Reply:
done